core/src/main/java/org/infinispan/jmx/AbstractJmxRegistration.java

```diff
@@ -18,65 +22,231 @@
+      if (mBeanServer == null) {
+         throw new IllegalStateException("MBean server not initialized");
+      }
+      try {
+         for (ResourceDMBean resourceDMBean : resourceDMBeans) {
+            if (resourceDMBean.getObjectName() != null) {
+               JmxUtil.unregisterMBean(resourceDMBean.getObjectName(), mBeanServer);
+            }
+         }
+      } catch (Exception e) {
+         throw new CacheException("Failure while unregistering MBeans", e);
       }
    }
 
-   protected void unregisterMBeans(Collection<ResourceDMBean> resourceDMBeans) {
-      if (mBeanServer != null) {
-         ComponentsJmxRegistration registrar = buildRegistrar();
-         registrar.unregisterMBeans(resourceDMBeans);
+   /**
+    * Creates an ObjectName based on given group and component name.
+    */
+   private ObjectName getObjectName(String groupName, String resourceName) throws MalformedObjectNameException {
+      if (groupName == null) {
+         throw new IllegalArgumentException("groupName cannot be null");
+      }
+      if (resourceName == null) {
+         throw new IllegalArgumentException("resourceName cannot be null");
       }
+      return new ObjectName(getDomain() + ":" + groupName + "," + COMPONENT_KEY + "=" + resourceName);
    }
 
-   protected Collection<ResourceDMBean> getResourceDMBeansFromComponents(Collection<ComponentRef<?>> components) {
-      Collection<ResourceDMBean> resourceDMBeans = new ArrayList<>(components.size());
+   /**
+    * Gathers all components from registry that have MBeanMetadata and creates ResourceDMBeans for them.
+    */
+   Collection<ResourceDMBean> getResourceDMBeansFromComponents() {
+      Collection<ComponentRef<?>> components = basicComponentRegistry.getRegisteredComponents();
+      Collection<ResourceDMBean> resourceDMBeans = new LinkedHashSet<>(components.size());   //todo make it ArrayList after confirming there are no duplicates
       for (ComponentRef<?> component : components) {
-         Object instance = component.wired();
-         if (instance != null) {
-            ResourceDMBean resourceDMBean = getResourceDMBean(instance, component.getName());
-            if (resourceDMBean != null) {
-               resourceDMBeans.add(resourceDMBean);
+         if (!component.isAlias()) {
+            Object instance = component.wired();
+            if (instance != null) {
+               ResourceDMBean resourceDMBean = getResourceDMBean(instance, component.getName());
+               if (resourceDMBean != null) {  // not all components have MBeanMetadata
+                  boolean added = resourceDMBeans.add(resourceDMBean);
+                  if (!added) {
+                     //todo [anistor] HACK to be removed! The Cache component is duplicated due to alias org.infinispan.Cache vs org.infinispan.AdvancedCache. hopefully no other duplication cases.
+                     throw new IllegalStateException("Found duplicate: " + resourceDMBean);
```

Comment:
Can this be updated now we have `ComponentRef.isAlias()` https://github.com/infinispan/infinispan/commit/f311d1a0281d61680f144b64b1e9b28d0ad146cd?

Reply:
I'll remove it. I thought I did that already. The TODO and the IllegalStateException was there to have at least a CI run to confirm the condition holds and the isAlias fix is indeed fixing this. Seems fine now.